File: Questão-11/Abstrato.go
package main

import(
	"fmt"
)

type Funcionario interface{
	CalcularSalario()
}
type BaseFuncionario struct{
	nome string
}

type FuncionarioHorista struct{
	BaseFuncionario
}
type FuncionarioAssalariado struct{
	BaseFuncionario
}

func (f FuncionarioHorista) CalcularSalario(ganhaporhora float64, horastrabalhadas int){
	var salario float64 = (ganhaporhora * float64(horastrabalhadas)) * 30
	fmt.Printf("\nO Funcionário %s ganha %.2f", f.nome, salario)
}

func (f FuncionarioAssalariado) CalcularSalario(ganhaporhora float64){
	var salario float64 = (ganhaporhora * 8) * 30
	fmt.Printf("\nO Funcionário %s ganha %.2f", f.nome, salario)
}




func main(){
	funcionario1 := FuncionarioHorista{BaseFuncionario: BaseFuncionario{nome:"João"}}
	funcionario2 := FuncionarioAssalariado{BaseFuncionario: BaseFuncionario{nome:"Fernando"}}

	funcionario1.CalcularSalario(20, 12)
	funcionario2.CalcularSalario(20)

}

File: Questão-11/Abstrato.py
from abc import ABC

class Funcionario(ABC):
    def __init__(self, nome):
        self.nome = nome
    
    def CalcularSalario(self):
        pass
    
class FuncionarioHorista(Funcionario):
    def CalcularSalario(self, horastrabalhadas, ganhaporhora):
        print(f"O Funcionario ganha {(horastrabalhadas * ganhaporhora) * 30}")
        
class FuncionarioAssalariado(Funcionario):
    def CalcularSalario(self, ganhaporhora):
        print(f"O Funcionario ganha {(ganhaporhora * 8) * 30}")
        

funcionario1 = FuncionarioAssalariado("João")
funcionario2 = FuncionarioHorista("Fernando")

funcionario1.CalcularSalario(20)
funcionario2.CalcularSalario(12, 20)